### PR TITLE
App review

### PR DESCRIPTION
**Pull request by Flou21**

### CODE REVIEW

**Comment by Flou21** [comment]
Warum funktioniert bootstrap nicht standardmäßig, nur mit cdn?

File: src/api/ApiHelper.tsx [hunk 1]
```diff
@@ -418,6 +420,70 @@ function initAPI(): API {
...
+        return new Promise((resolve, reject) => {
+            websocketHelper.sendRequest({
+                type: RequestType.GET_STRIPE_PRODUCTS,
+                data: null,
+                resolve: (products: any) => {
+                    getStripePrices().then((prices: Price[]) => {
+                        resolve(products.data.filter((product: any) =>
+                            product.active
+                        ).map((product: any) => {
+                            console.log(product);
```
Comment (matthias-luger) [inline src/api/ApiHelper.tsx:433]: remove console.log

File: src/api/ApiHelper.tsx [hunk 2]
```diff
@@ -418,6 +420,70 @@ function initAPI(): API {
...
+        })
+    }
+
+    let getStripePrices = (): Promise<Price[]> => {
+        return new Promise((resolve, reject) => {
+            websocketHelper.sendRequest({
+                type: RequestType.GET_STRIPE_PRICES,
+                data: null,
+                resolve: (prices: any) => {
+                    let mappedPrices = prices.data.map((price: any) => {
```
Comment (matthias-luger) [inline src/api/ApiHelper.tsx:458]: Extract functionality into ApiResponseParser

File: src/api/ApiHelper.tsx [hunk 3]
```diff
@@ -418,6 +420,70 @@ function initAPI(): API {
...
     }
 
+    let getStripeProducts = (): Promise<Product[]> => {
+        return new Promise((resolve, reject) => {
+            websocketHelper.sendRequest({
+                type: RequestType.GET_STRIPE_PRODUCTS,
+                data: null,
+                resolve: (products: any) => {
+                    getStripePrices().then((prices: Price[]) => {
+                        resolve(products.data.filter((product: any) =>
```
Comment (matthias-luger) [inline src/api/ApiHelper.tsx:430]: extract functionality into ResponseParser

File: src/api/WebsocketHelper.tsx [hunk 4]
```diff
@@ -20,6 +20,7 @@ function initWebsocket(): void {
     };
 
     let onWebsocketMessage = (e: MessageEvent): void => {
+        console.log({e});
```
Comment (matthias-luger) [inline src/api/WebsocketHelper.tsx:23]: remove console.log

File: src/components/Payment/Payment.tsx [hunk 5]
```diff
@@ -1,27 +1,74 @@
...
+  let [productListJsx, _setProductListJsx] = useState<JSX.Element>();
+
+  const setProductsListJsx = (jsx: JSX.Element) => {
+    if (productListJsx !== jsx) {
+      productListJsx = jsx;
+      _setProductListJsx(productListJsx);
+    }
+  }
+
+  const getProducts = async () => {
```
Comment (matthias-luger) [inline src/components/Payment/Payment.tsx:19]: remove async await

File: src/components/Payment/Payment.tsx [hunk 6]
```diff
@@ -1,27 +1,74 @@
...
+    if (!paymentProvider) {
+      paymentProvider = await availablePaymentProvider();
     }
-  };
+    return await paymentProvider.getProducts();
+  }
+
+  const getProductsJsx = async (): Promise<JSX.Element> => {
```
Comment (matthias-luger) [inline src/components/Payment/Payment.tsx:26]: Why are you doing it like this?
Just put the products into a state-variable.
Remove async await
Remove col-classNames (very large files)

File: src/components/Payment/Payment.tsx [hunk 7]
```diff
@@ -1,27 +1,74 @@
...
+              Buy
+            </Button>
+          </div>
+        </div>)
+        }
+      </div>
+    )
+  }
+
+  const roundToTwo = (param: number): number => Math.round(param * 10 ** 2) / 10 ** 2;
```
Comment (matthias-luger) [inline src/components/Payment/Payment.tsx:58]: better: Math.round((num + Number.EPSILON) * 100) / 100

File: src/components/Payment/Payment.tsx [hunk 8]
```diff
@@ -1,27 +1,74 @@
...
+
+  const onPay = async (product: Product) => {
+    paymentProvider.pay(product);
+  }
+
+  const setProductsJsx = async () => setProductsListJsx(await getProductsJsx())
+
+  useEffect(() => { setProductsJsx() }, []);
```
Comment (matthias-luger) [inline src/components/Payment/Payment.tsx:66]: use effect goes to the top of the file

File: src/components/Payment/Payment.tsx [hunk 9]
```diff
@@ -1,27 +1,74 @@
...
+    )
+  }
+
+  const roundToTwo = (param: number): number => Math.round(param * 10 ** 2) / 10 ** 2;
+
+  const onPay = async (product: Product) => {
+    paymentProvider.pay(product);
+  }
+
+  const setProductsJsx = async () => setProductsListJsx(await getProductsJsx())
```
Comment (matthias-luger) [inline src/components/Payment/Payment.tsx:64]: remove async await

File: src/components/Payment/Payment.tsx (hunk 6 above)
Comment (matthias-luger) [inline src/components/Payment/Payment.tsx:26]: JSX variables or functions go to the bottom of the file

File: src/components/Search/Search.tsx [hunk 10]
```diff
@@ -5,6 +5,7 @@ import './Search.css';
 import { useHistory } from "react-router-dom";
 import { convertTagToName } from '../../utils/Formatter';
 import NavBar from '../NavBar/NavBar';
+import Payment from '../Payment/Payment'
```
Comment (matthias-luger) [inline src/components/Search/Search.tsx:8]: remove unused import

File: src/components/SubscriptionList/SubscriptionList.tsx [hunk 11]
```diff
@@ -9,6 +9,7 @@ import { convertTagToName, numberWithThousandsSeperators } from '../../utils/For
 import NavBar from '../NavBar/NavBar';
 import { wasAlreadyLoggedIn } from '../../utils/GoogleUtils';
 import { toast } from 'react-toastify';
+import Payment from '../Payment/Payment';
```
Comment (matthias-luger) [inline src/components/SubscriptionList/SubscriptionList.tsx:12]: remove unused import

File: src/components/SubscriptionList/SubscriptionList.css [hunk 12]
```diff
@@ -1,4 +1,8 @@
 .subscription-list {
     width: 100%;
     font-size: medium;
+}
+
+.buy-prime-button {
```
Comment (matthias-luger) [inline src/components/SubscriptionList/SubscriptionList.css:6]: remove unused class

File: src/react-app-env.d.ts [hunk 13]
```diff
@@ -140,4 +143,41 @@ interface ApiResponse {
...
+interface Product {
+    itemId: string,
+    title: string,
+    description: string,
+    price: Price,
+    introductoryPrice: Price
+}
+
+interface Price {
+    productId: string|null,
```
Comment (matthias-luger) [inline src/react-app-env.d.ts:157]: format code

File: src/react-app-env.d.ts [hunk 14]
```diff
@@ -140,4 +143,41 @@ interface ApiResponse {
...
+    currency: string,
+    value: float
```
Comment (matthias-luger) [inline src/react-app-env.d.ts:159]: why float and not number

File: src/react-app-env.d.ts [hunk 15]
```diff
@@ -140,4 +143,41 @@ interface ApiResponse {
...
+}
+
+interface PaymentDetailsTotal {
+    label: string,
+    amount: PaymentDetailTotalAmount
+}
+
+interface PaymentDetailTotalAmount {
+    currency: string,
+    value: string
```
Comment (matthias-luger) [inline src/react-app-env.d.ts:182]: is type string for a value really correct?

File: src/utils/GoogleUtils.tsx [hunk 16]
```diff
@@ -1,3 +1,7 @@
 export function wasAlreadyLoggedIn(){
     return localStorage.getItem("googleId") !== null;
-}
\ No newline at end of file
+}
+
+export function getPackageName () {
```
Comment (matthias-luger) [inline src/utils/GoogleUtils.tsx:5]: why is this a function and not a constant

File: src/utils/Payment/GooglePlayProvider.tsx [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    data: {
+        sku: 'premium_1'
+    }
+}];
+
+export default class GooglePlayProvider extends AbstractPaymentProvider {
+
+    private digitalGoodsService: any;
+
+    public async getProducts(): Promise<Product[]> {
```
Comment (matthias-luger) [inline src/utils/Payment/GooglePlayProvider.tsx:24]: remove async await

File: src/utils/Payment/GooglePlayProvider.tsx [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+let paymentMethods: PaymentMethod[] = [{
+    supportedMethods: "https://play.google.com/billing",
+    data: {
+        sku: 'premium_1'
+    }
+}];
+
+export default class GooglePlayProvider extends AbstractPaymentProvider {
```
Comment (matthias-luger) [inline src/utils/Payment/GooglePlayProvider.tsx:20]: we use functional components

File: src/utils/Payment/AbstractPaymentProvider.tsx [hunk 19]
```diff
@@ -0,0 +1,11 @@
+import GooglePlayProvider from "./GooglePlayProvider";
+import StripePaymentProvider from "./StripePaymentProvider";
+
+export default abstract class AbstractPaymentProvider {
```
Comment (matthias-luger) [inline src/utils/Payment/AbstractPaymentProvider.tsx:4]: use functional components

File: src/utils/Payment/GooglePlayProvider.tsx [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+            await this.setDigitalGoodsService();
+        }
+        if (this.digitalGoodsService) {
+            let result = await this.digitalGoodsService.getDetails(['premium_30', 'premium_1', 'premium_3']);
+            return result;
+        }
+        return [];
+    }
+
+    public async pay(product: Product): Promise<Product> {
```
Comment (matthias-luger) [inline src/utils/Payment/GooglePlayProvider.tsx:35]: remove async await

File: src/utils/Payment/GooglePlayProvider.tsx [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+            await this.digitalGoodsService.acknowledge(token, 'onetime');
+            await paymentResponse.complete('success');
+            return product;
+        } else {
+            await paymentResponse.complete('fail');
+            return product;
+        }
+    }
+
+    public async checkIfPaymentIsPossible(): Promise<boolean> {
```
Comment (matthias-luger) [inline src/utils/Payment/GooglePlayProvider.tsx:51]: remove async await

File: src/utils/Payment/GooglePlayProvider.tsx [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+            return false;
+        }
+        return true;
+    }
+
+    private async validatePaymentToken(token: string, product: Product): Promise<boolean> {
+        return api.validatePaymentToken(token, product.itemId);
+    }
+
+    private async setDigitalGoodsService() {
```
Comment (matthias-luger) [inline src/utils/Payment/GooglePlayProvider.tsx:65]: remove async await

File: src/utils/Payment/GooglePlayProvider.tsx [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+        if (!window.PaymentRequest) {
+            return false;
+        }
+        if (!('getDigitalGoodsService' in window)) {
+            return false;
+        }
+        return true;
+    }
+
+    private async validatePaymentToken(token: string, product: Product): Promise<boolean> {
```
Comment (matthias-luger) [inline src/utils/Payment/GooglePlayProvider.tsx:61]: remove async await

File: src/utils/Payment/PaymentUtils.tsx [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+const paymentProviders = [
+    GooglePlayProvider,
+    StripePaymentProvider
+]
+
+const defaultPaymentProvider = StripePaymentProvider;
+
+let currentProvider: AbstractPaymentProvider|null = null;
+
+export default async function availablePaymentProvider(): Promise<AbstractPaymentProvider> {
```
Comment (matthias-luger) [inline src/utils/Payment/PaymentUtils.tsx:14]: function should be called PaymentUtils

File: src/utils/Payment/StripePaymentProvider.tsx [hunk 25]
```diff
@@ -0,0 +1,34 @@
+import AbstractPaymentProvider from "./AbstractPaymentProvider";
+import { loadStripe } from "@stripe/stripe-js";
+import api from '../../api/ApiHelper';
+
+export default class StripePaymentProvider extends AbstractPaymentProvider {
```
Comment (matthias-luger) [inline src/utils/Payment/StripePaymentProvider.tsx:5]: use functional components

File: src/utils/Payment/StripePaymentProvider.tsx [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+        return product;
+    }
+
+    public async checkIfPaymentIsPossible(): Promise<boolean> {
+        return true;
+    }
+
+
+    googleId = (): string | null => {
+        //TODO move this function
```
Comment (matthias-luger) [inline src/utils/Payment/StripePaymentProvider.tsx:31]: no TODO

File: src/utils/Payment/StripePaymentProvider.tsx [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+            "pk_test_51I6N5ZIIRKr1p7dQOGhRRigwIMqgZ3XnoBdbfezFNFgLiR9iaW2YzkRP9kAADCzxSOnqLeqKDVxglDh5uxvY28Dn00vAZR7wQ9"
+        );
+    }
+
+    public async getProducts(): Promise<Product[]> {
+        return await api.getStripeProducts();
+    }
+
+    public async pay(product: Product): Promise<Product> {
+        api.pay(this.stripePromise, product);
```
Comment (matthias-luger) [inline src/utils/Payment/StripePaymentProvider.tsx:21]: you forgot the await
remove the async await

**Review by matthias-luger — CHANGES_REQUESTED** [review]

**Review by matthias-luger — APPROVED** [review]

**Pull request merged**